### ISSUE TITLE
Topics and outline

### ISSUE THREAD

**Issue by shiffman** [processing/Processing-Hour-Of-Code#17]
This is mostly a note to myself that I want to go through the videos and outline the topics I presented in preparation of a "final" video shoot.  I also recall while making the video some questions came up, such as:
1. Should I explain `size()` right from the start, in the middle, or never?
2. Are we doing alpha with color or just RGB?
3. Are we including "else" with conditionals?  I think yes.

There may be more this is just what I recall.

**Comment by shiffman** [comment]
We now have a script here:

https://github.com/scottgarner/Processing-Hour-Of-Code/blob/gh-pages/Script.md

Questions:
- Shapes
  - My plan is to not mention size() until "interact" when setup() and draw() are introduced.  @scottgarner and I discussed pre-processing the code to completely ignore size() and always render the canvas with a fixed size.  I'll have to say somewhere that the size() is not adjustable for the purposes of this tutorial.
- Color
  - Should I demonstrate grayscale as a single value or just as 3 equal values?
  - Should we try to cover alpha / transparency?
  - Really I guess I'm asking if we should stray beyond 3 arguments only
- Questions
  - Should I include "else"?  I think so.

**Comment by REAS** [comment]
Shapes: OK, if possible.
Color: I think yes, a single value for gray, it makes sense to have "more values" for color. If you think the final example code doesn't need alpha, then yes, let's leave it out.
Question: Yes, I think "else" is good to include.

**Comment by scotthmurray** [comment]
Agreed with @reas on all points.

@shiffman, are you saying that size() will not be adjustable _until_ the Interact section? If so, then I just wouldn't mention size() at all until then.  If you're saying that size() will not be adjustable _ever_ in this hour, then I would just avoid mentioning size ever, as it's out of scope.

Depending on how much time we have in the videos, I might lean toward the latter option — just commit to a fixed size output panel (500x400) for the duration.

**Comment by shiffman** [comment]
@alignedleft yes, i am saying that size() will never be adjustable in this hour tutorial.  The reason I thought to mention it anyway at the moment of "interact" was that when I introduce setup() and draw() I have nothing to put in setup().  I guess I could leave it blank saying this is a simple program that is only drawing, and then reveal how setup() works when background() moves from draw() to setup()?  How does this sound?

**Comment by shiffman** [comment]
I could also introduce "void draw()" only first and not mention setup(), just talking about looping.  and then add "setup()" afterwards when I want to move background() there?

**Comment by scotthmurray** [comment]
Good question, as it seems odd to introduce setup() without putting anything in it.  Though we can't use draw() without it.

Your suggestions make sense.  If it helps, when you first introduce setup(), you could include a comment to indicate what happens there, something like:

```
void setup() {
  //This runs once when we start the sketch
}
```

Then, after introducing draw(), you could circle back to setup() and add background() and whatever else.  Is that useful?

**Comment by shiffman** [comment]
Ok, so I rewatched all my test videos this morning and took notes and made the script outline a lot more detailed.  Would love any feedback.

https://github.com/scottgarner/Processing-Hour-Of-Code/blob/gh-pages/Script.md

I could also use thoughts on the following questions (which are also on the script).

Shapes
1. Question: I switched to introducing rect() first then ellipse(), how does that seem?
2. Question: Is the language "command" ok ellipse(), rect() -- should I say also say "function"?
3. Question: what are some other ideas for exercise suggestions?  Design a face, alien, logo, building,. . . ?
4. Question: should I suggest a specific amount of time for the exercise?

Color
1. Question: I think we should skip transparency, this feels like plenty? we can always add it to the hints???
2. Question: @scottgarner, the following is possible right? "I'm going to put an example below that you could use to begin this exercise or you can revert back to your shapes exercise (that you completed after watching video #1).

Interact
- Question: I still can't decide about size.  I think we should either:
  - never show it and leave setup() blank until I move background() there
  - show it and allow them to change the canvas size() but warn that it may no longer fit into the tutorial site if they make it too big

Also, happy to take suggestions / thoughts about what to say for "What is programming?" and the wrap-up video.

**Comment by scottgarner** [comment]
Shapes:
1. That's tricky since the first shape is also the first introduction to the coordinate system. I'd lean towards whatever you think best represents that. In a certain way, ellipse is nice because the middle is right where you'd expect it to be, but I'm undecided.
2. Somehow "command" does seem like more beginner-friendly language, though I'm not sure why I think that.
3. Maybe a reference to mid-century modern art? Rothko or Albers?
4. I've been wondering about the time thing since the beginning. I don't have a good answer, but my instinct is that just telling them to play around for a few minutes is better than saying you have exactly 5 minutes to experiment or something like that.

Color:
1. Skipping transparency seems wise. Especially since it's not supported in the color picker, though it potentially could be with more work.
2. We haven't discussed save and restore very much in a practical way. It's certainly possible, but as of now we have no plan of how the interaction would work. I'm wary of it in general because there are so many possible use cases we'd have to account for. Unless we can come up with a concrete plan of how this should work, I'm not very comfortable including it.

We could automatically save whatever is in the editor when moving to another lesson, but then we at least need buttons to "Reload my own code" and to "Reload Dan's example". This also implies that saving only happens once when switching lessons and that there's no general "save" and "load" functionality, which could get really complicated. But what happens if you go back a lesson? What happens if you come be back later and start at video 3? There won't be any code to restore. I have the feeling that in the process of implementing something like this, a dozen similar roadblocks would arise.

Sorry, I'm kind of thinking as I type. Here's my overall take at this moment:

I don't think this should be mentioned in the video. If only because of the problem I mentioned above of coming back to the site and starting at a later video. If we really need it, I think it could be implemented as a button in the hints section where we have the ability to hide it if there's no code to load.

Interact:
1. Changing size has other implications, such as the need for dynamic rulers. Not a huge issue, but probably a few more hours of work.

Speaking broadly, there are still a lot of functional chores to do (social sharing, etc.), not to mention the work of adding, annotating and styling around the new videos. My guess is all of these tasks currently add up to around 25 hours of work. After this week I won't be able to dedicate the same amount of time I have up to this point, so adding major changes now worries me.

**Comment by REAS** [comment]
[I'm writing quickly because there's too much material to cover and craft the language, please forgive if it seems rude.]

Shapes:
1. I think rect() first is good. The upper-left corner is easy to see and it makes sense.
2. We use the word "function" in all Processing documentation and it's the term in p5.js. I would mention that it's different from a math function, which many of these students are likely struggling with.
3. Can we put an image up in the "hint" area and ask them to draw it. We can mention they can try to draw this "creature" or invent one
4. I think we need to because we're on a one-hour clock. Yes? No?

Color:
1. Yes, leaving out alpha for color is good.

Saving:
1. I agree with Scott. I think we have our plate full and if we try to implement "save" there will be so many special cases that we'll likely be setting ourselves up to not meet the expectations.
2. Can we use the "share" button or rename that and invite people to save for themselves? For example at the end of an exercise you can tell them if they want to save their code to cut and paste it somewhere or to hit the share button and then save that link.
3. Or we can simply tell people that we won't be saving work until that last exercise. This is just practice, then in the final tutorial we can have the share link and people can click there to get a URL?

Interact: 
1. I prefer to allow people to change the size in the tutorial, but we'll add it only after you start with setup(). They can make it too large or very small if they like. Having size() at the beginning of setup() is essential to Processing and I think it should be included if possible.

**Comment by scottgarner** [comment]
I think "share" as save is a really good solution overall. Currently you end up with a URL you can return to on the display page. Behind that is a GIST, though, that could be downloaded or whatever. That just means adding a few simple additions to the display page to view/download code.

I think maybe this is where an important distinction needs to be made. I would say at this point if they want to continue with that saved code from the lessons, visitors should be encouraged to take that code to Processing instead of continuing in our admittedly limited environment.

I guess that still doesn't quite solve the problem of Dan's basic notion of carrying code from lesson to lesson, but I think the absolute simplest approach is Casey's thought of only offering share/save a the end.

**Comment by REAS** [comment]
I think the cleanest, simplest solution is to not try to save code from lesson to lesson. 

Dan can tell them to cut and paste their code at the end of each or to use "share" to get a URL to their code. Then we can start fresh at each step. It's not ideal, but it seems like a good way to do it given that we don't want people to have "accounts" and to build an entire save/load mechanism.

**Comment by scottgarner** [comment]
I wonder if it would help to change the language from "share" to something like "snapshot" with the explanation that they can reference that code later or share it on social media, but not necessarily bring it back into the editor unless they cut and paste. That feels pretty clumsy, unfortunately. 

There's the possibility of storing snapshots as cookies so that they _could_ be reloaded directly into the editor, but that again starts to go down the road of complexity.

**Comment by shiffman** [comment]
I agree that we should not implement a "save" and "restore" due to complexity.  i also do not like asking people to cut and paste as I think this tutorial should work entirely in the browser.    I do, however, feel like people will want to continue from where they left off.  I'm not so worried about those coming back to the middle of the tutorial.  I think we should make it work primarily for the full tutorial experience and if you jump into the middle it's fine if the experience isn't seamless and there's nothing for you to load.

All that said, my preference would be for:

a) an option to restore back to your code state after video N when video N+1 finishes.
b) this new "snapshot" idea where I say if you want to save your code before the next lesson make a "snapshot" and you can get it back by copy/pasting.

I would err on the side of whatever is easiest and the least time  consuming for Scott.

Is this a helpful summary?

**Comment by scottgarner** [comment]
Dan, are you around today to chat about this?

**Comment by scottgarner** [comment]
Skype would be fine, too, it just might be helpful to talk it through for a moment.

**Comment by shiffman** [comment]
am not at ITP but just sent you a contact request on skype.  am free this afternoon.

**Comment by scottgarner** [comment]
Great. I have a meeting until three or so. I think I have some ideas of how we can solve this with your language in the video rather than a lot of code.

**Comment by shiffman** [comment]
@REAS if you happen to be free when @scottgarner and i skype, i'll conference you in.

**Comment by shiffman** [comment]
ah, i'm busy between 2:30 and 4.  Could we say 4:30?

**Comment by shiffman** [comment]
also @scottgarner if you are around ITP tomorrow we could discuss and sort this out before/during the video shoot (planned for 11am - 5pm)

**Comment by scottgarner** [comment]
I'm happy to talk about it tomorrow at 10am if that's better. Or this afternoon/evening if Casey would like to join.

**Comment by scotthmurray** [comment]
For the discussion around "saving" code between lessons:  A compromise could be to "save" the editor contents at the end of each lesson (or arbitrarily, anytime?), but it would only be "saved" in memory (not to anything on the back-end, and not shareable).  

For example, Sketchpad does this, where it saves the code state every time you hit "run".  But it saves on the back-end, too.

I'm picturing a super simple interface to access prior "saved" versions, like tabs/buttons beneath the editor panel: v1, v2, v3, v4…  You could click one to restore the saved version, overwriting your current one.

Yet I think this is more a UI issue than a technical one.  I'm not sure how to make the UI for this make sense without being confusing or misleading.

…which I guess brings me back to vote for not saving work at all (except to "share" at the very end).

**Comment by shiffman** [comment]
thanks for the thoughts @alignedleft!  @scottgarner let's try to talk late this afternoon if we can and tomorrow will be a backup.

**Comment by scottgarner** [comment]
A run cache like that is pretty simple since I already cache every run—it would just mean pushing to an array instead of a single variable. However, the thing that complicates it in our case is that sometimes the user is running things and sometimes Dan is and we do want to be able to revert to his code sometimes, too. And yes, the UI is really the big issue.

Here are my current thoughts. Dan, let's plan to quickly chat on Skype at 4:30 and I'll be on the floor in the morning, too.

1) I think everyone is agreed that a robust save/load system is out of scope, so lets call that settled.

2) That leaves us with two use cases to account for, the first of which is Dan's desire to carry code from one lesson to the next. I see the value in that and I think I have a way we can kind of talk around issues of state, like if someone jumps in at lesson 3, so there's no code for lesson 2. Basically I'll just save every run to an array indexed by the proper lesson, so we can always look at lesson - 1 and load it if it exists. We just need a few more buttons and tooltips will help.

3) The second use case is our share/snapshot system. However we phrase it, I think the understanding here is that once you use it, you get a link and that code is now out of the lesson system and into the world. 

We can discuss what kinds of interactions might be nice to have on the display page, like the ability to view and download the code, but I think once you use this function, the expectation shouldn't be that it comes back into the hello editor. Maybe you just view it, maybe you can download a PDE, but I think that can all happen on the display page.

We can also discuss whether you get the share button all the time or only at the end. I think if we establish the above, it could work either way. My instinct is that it's cleaner to keep it at the end, though.

**Comment by shiffman** [comment]
First thank you to everyone for all the back and forth and great thoughts and ideas today!  I think we are on a good path.  I worked a bit more on the scripts and brought in the great work casey did to put together a sampling of interesting Processing work.

Hello Script: https://github.com/scottgarner/Processing-Hour-Of-Code/blob/gh-pages/Script-Hello-Intro.md

Tutorials Script: https://github.com/scottgarner/Processing-Hour-Of-Code/blob/gh-pages/Script-tutorials.md

Please feel free to look these over and make any comments.
- @REAS -- do we want to try to slip in the eels last minute since we lost the one other animation project or not bother?
- The main thing I added since this morning is a little section of notes on "What is programming" which i will cover at the end of the art and design intro and leading into coordinate systems.  would love any feedback.  I'll paste those notes here:

### Programming and the visual arts
- See [Hello Intro Script](https://github.com/scottgarner/Processing-Hour-Of-Code/blob/gh-pages/Script-Hello-Intro.md)

### What is Programming
- writing a set of instructions for your computer to follow
- like telling your friend how to cook an omlette or walk from one place to another
- only the computer has its own special language, highly specific and with odd-looking syntax.  what may look strange now can become second nature just like speaking in your native tongue.
- programming is used for everything from solving scientific problems to building web sites to creating art
- here we are going to learn the basics of programming -- what is that syntax? how do we write these instructions? -- in the context of drawing, animation, and interactivity.
- Even though we're just going to be drawing and manipulating simple shapes on the screen, the fundamental concepts we'll learn are the same concepts used by programmers all over the world to build every piece of software you've ever used or seen.  and once you learn these fundamentals, anything you can imagine, you can program.
- let's begin. . . .

**Comment by REAS** [comment]
What is Programming
- Is "syntax" jargon? Can we say "unfamiliar symbols" in place of "odd-looking jargon"?

**Comment by REAS** [comment]
With the intro script, the animation and installation/motion/dance projects are in the first part and we have four there. We lost one project from the graphic design/print section. If we add the eels, that would put 5 in the first section, 2 in the second, and three in the last.

**Comment by shiffman** [comment]
Well, we just shot all the videos so this issue is now CLOSED!!!   We'll be adding a series of new issues based on all the weird things I said that the web site would do while I was rambling.

I will also just note that for the first paragraph and last paragraph of the "hello" script, instead of reading it, I just said things off the cuff which felt more natural.  The read parts will mostly be voiceover.  We felt this was less weird.  I also added some things like "hello, welcome to what this is, my name is. . . blah blah"
